lib/simple-seeder.js

```diff
@@ -0,0 +1,221 @@
+const Hyperbee = require('hyperbee')
+const Hyperdrive = require('hyperdrive')
+const Seeders = require('@hyperswarm/seeders')
+const speedometer = require('speedometer')
+const HypercoreId = require('hypercore-id-encoding')
+const mutexify = require('mutexify/promise')
+
+module.exports = class SimpleSeeder {
+  constructor (store, swarm, opts = {}) {
+    this.store = store
+    this.swarm = swarm
+
+    this.backup = opts.backup === true
+
+    this.lock = mutexify()
+    this.resources = new Map()
+
+    this.destroying = false
+  }
+
+  async add (key, opts) {
+    const release = await this.lock()
+    try {
+      return await this._add(key, opts)
+    } finally {
+      release()
+    }
+  }
+
+  async _add (key, opts) {
+    if (this.destroying) return
+    if (this.has(key)) throw new Error('Resource already added')
+
+    const publicKey = HypercoreId.decode(key)
+    const id = HypercoreId.encode(publicKey)
+
+    // Maybe all this belongs to a "Resource" class, including the internals _createCore/Bee/etc
+    const info = {
+      key: id,
+      type: opts.type,
+      instance: null,
+      description: opts.description || null,
+      seeders: null,
+      discovery: null,
+      blocks: { down: speedometer(), up: speedometer() },
+      network: { down: speedometer(), up: speedometer() },
+      external: opts.external !== false
+    }
+
+    if (opts.type === 'core') info.instance = await this._createCore(publicKey, info)
+    else if (opts.type === 'bee') info.instance = await this._createBee(publicKey, info)
+    else if (opts.type === 'drive') info.instance = await this._createDrive(publicKey, info, opts)
+    else if (opts.type === 'list') info.instance = await this._createBee(publicKey, info, { keyEncoding: 'utf-8', valueEncoding: 'json' })
+    else if (opts.type === 'seeders') {
+      info.type = 'drive'
+      info.instance = await this._createDrive(publicKey, info, { seeders: true })
+    } else throw new Error('Type is required')
+
+    this.resources.set(id, info)
+
+    return info
+  }
+
+  async put (key, opts) {
+    const release = await this.lock()
+    try {
+      return await this._put(key, opts)
+    } finally {
+      release()
+    }
+  }
+
+  async _put (key, opts) {
+    if (!this.has(key)) return this._add(key, opts)
+
+    const info = this.get(key)
+
+    if (info.type !== opts.type) {
+      await this._remove(key)
+      return this._add(key, opts)
+    }
+
+    if (!!info.seeders !== opts.seeders) {
```

Comment:
The if-else branching and the conditions here are hard to follow. Can this be flattened a bit more? For example, if you put `info.description = opts.description || null` first, you can just early return I think:

`    if (!!info.seeders === opts.seeders) return info`

 Also a bit confused by the underlying logic: when does it happen that info.seeders exists, and opts.seeders is not set, meaning we want to remove it?

Reply:
`await tracker.put(key, { type: 'drive', seeders: true })`
`await tracker.put(key, { type: 'drive', seeders: false })`

Reply:
`!!info.seeders !== opts.seeders` looks complicated I know but it doesn't get more complicated than that, it just checks that they're different in boolean values

(`info.seeders` is null or an object, and `opts.seeders` a bool)

Similar as the check above `info.type !== opts.type`

Reply:
It basically toggles the seeders